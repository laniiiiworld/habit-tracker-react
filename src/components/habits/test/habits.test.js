import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Habits from '../habits';

describe('Habits Component', () => {
  const habits = [
    { id: 1, name: 'Reading', count: 2 },
    { id: 2, name: 'Running', count: 0 },
  ];
  let HabitsComponent;
  let onAdd;
  let onIncrement;
  let onDecrement;
  let onDelete;
  let onReset;

  beforeEach(() => {
    onAdd = jest.fn();
    onIncrement = jest.fn();
    onDecrement = jest.fn();
    onDelete = jest.fn();
    onReset = jest.fn();
    HabitsComponent = (
      <Habits //
        habits={habits}
        onAdd={onAdd}
        onIncrement={onIncrement}
        onDecrement={onDecrement}
        onDelete={onDelete}
        onReset={onReset}
      />
    );
  });

  it('renders', () => {
    const component = renderer.create(HabitsComponent);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('Button click', () => {
    beforeEach(() => {
      render(HabitsComponent);
    });

    it('calls onAdd when clicking the "add" button', () => {
      const input = screen.getByPlaceholderText('Habit');
      const button = screen.getByText('Add');
      const newHabit = 'New Habit';
      userEvent.type(input, newHabit);
      userEvent.click(button);
      expect(onAdd).toHaveBeenCalledWith(newHabit);
    });

    it('calls onIncrement when clicking the "increment" button', () => {
      const button = screen.getAllByTitle('increment')[0];
      userEvent.click(button);
      expect(onIncrement).toHaveBeenCalledWith(habits[0]);
    });

    it('calls onDecrement when clicking the "decrement" button', () => {
      const button = screen.getAllByTitle('decrement')[0];
      userEvent.click(button);
      expect(onDecrement).toHaveBeenCalledWith(habits[0]);
    });

    it('calls onDelete when clicking the "delete" button', () => {
      const button = screen.getAllByTitle('delete')[0];
      userEvent.click(button);
      expect(onDelete).toHaveBeenCalledWith(habits[0]);
    });

    it('calls onReset when clicking the "Reset All" button', () => {
      const button = screen.getByText('Reset All');
      userEvent.click(button);
      expect(onReset).toHaveBeenCalledTimes(1);
    });
  });
});
