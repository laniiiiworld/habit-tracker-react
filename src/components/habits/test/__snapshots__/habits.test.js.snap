// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habits Component renders 1`] = `
Array [
  <form
    className="add-form"
    onSubmit={[Function]}
  >
    <input
      className="add-input"
      placeholder="Habit"
      type="text"
    />
    <button
      className="button add-button"
    >
      Add
    </button>
  </form>,
  <ul
    className="habits"
  >
    <li
      className="habit"
    >
      <span
        className="habit-name align-left"
        data-testid="habit-name"
      >
        Reading
      </span>
      <span
        className="align-right"
      >
        <span
          className="count habit-count"
          data-testid="habit-count"
        >
          2
        </span>
        <button
          className="habit-button habit-increase"
          onClick={[Function]}
          title="increment"
        >
          <i
            className="fa-solid fa-square-plus"
          />
        </button>
        <button
          className="habit-button habit-decrease"
          onClick={[Function]}
          title="decrement"
        >
          <i
            className="fa-solid fa-square-minus"
          />
        </button>
        <button
          className="habit-button habit-delete"
          onClick={[Function]}
          title="delete"
        >
          <i
            className="fa-solid fa-trash-can"
          />
        </button>
      </span>
    </li>
    <li
      className="habit"
    >
      <span
        className="habit-name align-left"
        data-testid="habit-name"
      >
        Running
      </span>
      <span
        className="align-right"
      >
        <span
          className="count habit-count"
          data-testid="habit-count"
        >
          0
        </span>
        <button
          className="habit-button habit-increase"
          onClick={[Function]}
          title="increment"
        >
          <i
            className="fa-solid fa-square-plus"
          />
        </button>
        <button
          className="habit-button habit-decrease"
          onClick={[Function]}
          title="decrement"
        >
          <i
            className="fa-solid fa-square-minus"
          />
        </button>
        <button
          className="habit-button habit-delete"
          onClick={[Function]}
          title="delete"
        >
          <i
            className="fa-solid fa-trash-can"
          />
        </button>
      </span>
    </li>
  </ul>,
  <button
    className="button reset-button"
    onClick={[MockFunction]}
  >
    Reset All
  </button>,
]
`;
